modules/python/python-stats.c

```diff
@@ -0,0 +1,175 @@
+
+typedef struct _PyStats
+{
+  PyObject_HEAD
+  StatsCounterItem *counter;
+  const char *stats_name;
+} PyStats;
+
+static PyTypeObject py_stats_type;
+
+PyObject *
+py_stats_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+  const gchar *name;
+  PyStats *self = (PyStats *) type->tp_alloc(type, 0);
+
+  if (self != NULL)
+    {
+      if (!PyArg_ParseTuple(args, "s", &name))
+        {
+          return NULL;
+        }
+      self->stats_name = g_strdup(name);
+
+      stats_lock();
+
+      StatsClusterKey sc_key;
+      stats_cluster_single_key_set(&sc_key, SCS_PYTHON, self->stats_name, NULL);
+      stats_register_counter(0, &sc_key, SC_TYPE_SINGLE_VALUE, &(self->counter));
+
+      stats_unlock();
+    }
+  return (PyObject *) self;
+};
+
+static void
+py_stats_free(PyStats *self)
+{
+  stats_lock();
+  StatsClusterKey sc_key;
+  stats_cluster_single_key_set(&sc_key, SCS_PYTHON, self->stats_name, NULL);
+  stats_unregister_counter(&sc_key, SC_TYPE_SINGLE_VALUE, &(self->counter));
+  stats_unlock();
+
+  g_free( (void *) self->stats_name);
+
+  PyObject_Del(self);
+}
+
+static PyObject *
+py_stats_increase(PyStats *self, PyObject *unused)
+{
+  stats_counter_inc(self->counter);
+  Py_RETURN_NONE;
+}
+
+static PyObject *
+py_stats_decrease(PyStats *self, PyObject *unused)
+{
+  stats_counter_dec(self->counter);
+  Py_RETURN_NONE;
+}
+
+static PyObject *
+py_stats_add(PyStats *self, PyObject *args)
+{
+  gssize amount;
+  if (!PyArg_ParseTuple(args,"n", &amount))
+    {
+      return NULL;
+    }
+  stats_counter_add(self->counter, amount);
+  Py_RETURN_NONE;
+}
+
+static PyObject *
+py_stats_sub(PyStats *self, PyObject *args)
+{
+  gssize amount;
+  if (!PyArg_ParseTuple(args,"n", &amount))
+    {
+      return NULL;
+    }
+
+  stats_counter_sub(self->counter, amount);
+  Py_RETURN_NONE;
+}
+
+static PyMethodDef py_stats_methods[] =
+{
+  {
+    "increase", (PyCFunction)py_stats_increase,
+    METH_NOARGS, "Increase counter"
+  },
+  {
+    "decrease", (PyCFunction)py_stats_decrease,
+    METH_NOARGS, "Decrease counter"
+  },
+  {
+    "add", (PyCFunction)py_stats_add,
+    METH_VARARGS, "Increase counter with amount"
+  },
+  {
+    "sub", (PyCFunction)py_stats_sub,
+    METH_VARARGS, "Decrease counter with amount"
+  },
+  {NULL}
+};
+
+static PyTypeObject py_stats_type =
+{
+  PyVarObject_HEAD_INIT(&PyType_Type, 0)
+  .tp_name = "StatsCounter",
+  .tp_basicsize = sizeof(PyStats),
+  .tp_dealloc = (destructor) py_stats_free,
+  .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
+  .tp_doc = "StatsCounter representing a counter in syslog-ng statistics",
+  .tp_new = py_stats_new,
+  .tp_methods = py_stats_methods,
+  .tp_str = NULL,
+  .tp_as_mapping = NULL,
+  0,
+};
+
+static PyMethodDef stats_module_methods[] =
+{
+  {NULL}  /* Sentinel */
+};
+
+
+PyMODINIT_FUNC
+PyInit_stats(void)
+{
+  PyObject *m;
+  if (PyType_Ready(&py_stats_type) < 0)
+    return;
+
+  m = Py_InitModule3("sngstats", stats_module_methods,
```

Comment:
This does not compile with python3
modules/python/libmod-python.so: undefined reference to `Py_InitModule3'

Reply:
Hey! Thanks for checking! Will fix it on the weekend!

Reply:
Fixed the python3 issue, tried with python 2.7 and 3.5, both worked